python/mxnet/gluon/data/dataloader.py

```diff
@@ -267,20 +316,128 @@ def __next__(self):
                 return batch
 
     def next(self):
+        """Compatible portal for __next__ in python2.
+
+        Returns
+        -------
+        type
+            Description of returned object.
+
+        """
         return self.__next__()
 
     def __iter__(self):
+        """Requiring iterator will reset current instance, but keep all workers
+        alive, thus save re-init time of forking processes.
+
+        Returns
+        -------
+        iterator
+            Iterator of self.
+
+        """
+        self._reset()
         return self
```

Comment:
What happens if users used the dataloader and created two iterators? Does the 2nd iterator reset the state for the first one that is running? For example:
```
>>> a = [1,2,3]
>>> for x in a:
...     for y in a:
...         print(x,y)
...
(1, 1)
(1, 2)
(1, 3)
(2, 1)
(2, 2)
(2, 3)
(3, 1)
(3, 2)
(3, 3)
```

Reply:
iterators from the same dataloader are two separate instances, they don't interfere with each other, however, for explicitly cached iterator, it will break

```
a = [1,2,3]
it = iter(a)
for x in it:
    for y in it:
        # this will not work